src/main/scala/edu/cmu/cs/obsidian/codegen/CodeGenYul.scala

```diff
@@ -0,0 +1,129 @@
+package edu.cmu.cs.obsidian.codegen
+
+import java.io.{File, FileReader, PrintWriter}
+import java.nio.file.{Path, Paths}
+
+import com.github.mustachejava.DefaultMustacheFactory
+import edu.cmu.cs.obsidian.CompilerOptions
+import edu.cmu.cs.obsidian.parser._
+import edu.cmu.cs.obsidian.codegen.YulAST
+
+import edu.cmu.cs.obsidian.Main.{findMainContract, findMainContractName}
+import com.github.mustachejava.Mustache
+import com.github.mustachejava.MustacheFactory
+
+import scala.collection.immutable.Map
+
+
+// need some table remembering field index in storage
+object CodeGenYul extends CodeGenerator {
+
+    // TODO improve this temporary symbol table
+    var temp_symbol_table: Map[String, Int] = Map() // map from field identifiers to index in storage
+    var temp_table_idx = 0
+
+
+    def gen(filename: String, srcDir: Path, outputPath: Path, protoDir: Path,
+            options: CompilerOptions, checkedTable: SymbolTable, transformedTable: SymbolTable): Boolean = {
+
+        val ast = checkedTable.ast
+        val mainContractOption = findMainContract(ast)
+        if (mainContractOption.isEmpty) {
+            throw new RuntimeException("No main contract found")
+        }
+
+        val mainName = findMainContractName(ast)
+
+        val finalOutputPath = options.outputPath match {
+            case Some(p) =>
+                Paths.get(p).resolve(mainName)
+            case None =>
+                Paths.get(mainName)
+        }
+
+        // temporary code, not designed to be recursive
+        val mainContract = mainContractOption.get
+        mainContract match {
+            case obsContract: ObsidianContractImpl => {
+                val translated_obj =  translateContract(obsContract)
+                yulString(translated_obj, finalOutputPath)
+            }
+            case _ => None
+        }
+//        val ast_seq =  translateProgram(ast)
+        return true
+    }
+
+    def translateProgram(program: Program): Unit = {
+        // ignore imports for now
+        for (c <- program.contracts) {
+            c match {
+                case obsContract: ObsidianContractImpl => translateContract(obsContract)
+                case javaContract: JavaFFIContractImpl => None
+            }
+        }
+    }
+
+    def translateContract(contract: ObsidianContractImpl): Object = {
+        // create runtime obj
+        var subobjects: Seq[Object] = Seq()
+        val runtime_name = contract.name + "_deployed"
+        val runtime_obj = Object(runtime_name, Code(Block(Seq())), Seq(), Seq())
+        subobjects = runtime_obj +: subobjects
+
+        var statement_seq: Seq[YulStatement] = Seq()
+        // initialization
```

Comment:
Delete commented-out code; you can stash it locally and add it back when you're ready.